### ISSUE TITLE
borrow sealing doesn't work for borrows within objects

### ISSUE THREAD

**Issue by alaviss** [nim-works/nimskull#1456]
### Example

```nim
{.experimental: "views".}
type
  Mint = object
    data: int

  Option = object
    value: lent Mint

proc `=copy`(dst: var Mint, src: Mint) {.error.}

proc some(x: Mint): Option =
  Option(value: x)

proc main() =
  var x = Mint(data: 10)
  let optlent = some(x)

  echo $optlent.value
  x = Mint(data: 12)
  echo $optlent.value

main()
```

### Actual Output

Compiles and run

### Expected Output

Error that `x` cannot be modified because `optlent` is holding a live borrow.

**Comment by saem** [comment]
Sorry if this is a silly question, but is `var x` being treated as `var x: var Mint`? So it's not really a true/independent location, but it's a view into `optLen.x`? Also is `optLen` still a live borrow because of the second `echo` meaning that `optLen` still has a read against it?

Did I get all those right?

**Comment by alaviss** [comment]
Nope, what's happening right now is that the borrow checker is not aware that `optlent` has borrowed `x`, so none of the constraints are applied.

**Comment by saem** [comment]
> Nope, what's happening right now is that the borrow checker is not aware that `optlent` has borrowed `x`, so none of the constraints are applied.

Sorry, I phrased my line of questioning incorrectly, I meant, "Is this what's meant to be happening ... <line of questions>"

**Comment by alaviss** [comment]
Then no, for this, `x` is the "owning" location of `Mint`, and `optlent` is a view into `x`.

Since `optlent` is referred to by the `echo` after `x = `, it's considered to be alive at the point of `x = `, which meant that the location `x` should remain sealed from mutations.
